src/adt/adt.py

```diff
@@ -76,95 +70,94 @@ def basic_check(i, name, argname, typ, indent="    "):
-            opt_check(i, f.name, f"arg_{i}", f.type) if f.opt else
-            basic_check(i, f.name, f"arg_{i}", f.type)
-            for i, f in enumerate(fields)
-        ])
-        assign = '\n    '.join([
-            f"self.{f.name} = arg_{i}"
-            for i, f in enumerate(fields)
-        ])
-        if len(fields) == 0:
+        nameargs = ", ".join([f.name for i, f in enumerate(fields)])
+        argstr = ", ".join([f"arg_{i}" for i, f in enumerate(fields)])
+        checks = "\n".join(
+            [
+                seq_check(i, f.name, f"arg_{i}", f.type)
+                if f.seq
+                else (
+                    opt_check(i, f.name, f"arg_{i}", f.type)
+                    if f.opt
+                    else basic_check(i, f.name, f"arg_{i}", f.type)
+                )
+                for i, f in enumerate(fields)
+            ]
+        )
+        assign = "\n    ".join(
+            [f"self.{f.name} = arg_{i}" for i, f in enumerate(fields)]
+        )
+        if not fields:
             checks = "    pass"
             assign = "pass"
 
-        exec_out = {'Err': Err, 'CHK': CHK}
-        exec_str = (f"def {C_name}_init_inner(self,{argstr}):\n"
-                    f"{checks}\n"
-                    f"    {assign}\n"
-                    f"def {C_name}_init(self,{nameargs}):\n"
-                    f"    {C_name}_init_inner(self,{nameargs})")
+        exec_out = {"Err": Err, "CHK": CHK}
+        exec_str = (
+            f"def {C_name}_init_inner(self,{argstr}):\n"
+            f"{checks}\n"
+            f"    {assign}\n"
+            f"def {C_name}_init(self,{nameargs}):\n"
+            f"    {C_name}_init_inner(self,{nameargs})"
+        )
         # un-comment this line to see what's
         # really going on
         # print(exec_str)
         exec(exec_str, exec_out)
-        return exec_out[C_name + '_init']
+        return exec_out[C_name + "_init"]
 
     def create_reprfn(C_name, fields):
-        prints = ','.join([
-            f"{f.name}={{repr(self.{f.name})}}"
-            for f in fields
-        ])
-        exec_out = {'Err': Err}
-        exec_str = (f"def {C_name}_repr(self):"
-                    f"\n    return f\"{C_name}({prints})\"")
+        prints = ",".join([f"{f.name}={{repr(self.{f.name})}}" for f in fields])
+        exec_out = {"Err": Err}
+        exec_str = f"def {C_name}_repr(self):" f'\n    return f"{C_name}({prints})"'
         # un-comment this line to see what's
         # really going on
         # print(exec_str)
         exec(exec_str, exec_out)
-        return exec_out[C_name + '_repr']
+        return exec_out[C_name + "_repr"]
 
     def create_eqfn(C_name, fields):
-        compares = ' and '.join(
-            ['type(self) == type(o)'] +
-            [f"(self.{f.name} == o.{f.name})" for f in fields]
+        compares = " and ".join(
+            ["type(self) == type(o)"]
+            + [f"(self.{f.name} == o.{f.name})" for f in fields]
```

Comment:
As mentioned below, should this be `is` instead of `==`?

Reply:
`is` for the `type` comparison, most likely.